File: app/dtos/community.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel

from app.dtos.base import BaseSerializerModel
from app.models.community import InfoCategory, PostCategory, QuizCategory, QuizOption, ReportReason, ReportTargetType


class PostListItem(BaseSerializerModel):
    id: int
    title: str
    category: PostCategory
    info_category: InfoCategory | None = None
    is_pinned: bool
    view_count: int
    comment_count: int
    like_count: int = 0
    author_id: UUID
    author_nickname: str | None
    created_at: datetime


class PostDetailResponse(PostListItem):
    content: str
    updated_at: datetime
    is_liked: bool = False


class PostListResponse(BaseModel):
    items: list[PostListItem]
    total: int
    page: int
    size: int


class PostCreateRequest(BaseModel):
    title: str
    content: str
    category: PostCategory
    info_category: InfoCategory | None = None


class PostUpdateRequest(BaseModel):
    title: str | None = None
    content: str | None = None
    category: PostCategory | None = None
    info_category: InfoCategory | None = None


class CommentResponse(BaseSerializerModel):
    id: int
    content: str
    author_id: UUID
    author_nickname: str | None
    parent_id: int | None
    created_at: datetime
    updated_at: datetime
    like_count: int = 0
    is_liked: bool = False
    replies: list["CommentResponse"] = []


CommentResponse.model_rebuild()


class CommentCreateRequest(BaseModel):
    content: str
    parent_id: int | None = None


class CommentUpdateRequest(BaseModel):
    content: str


class LikeResponse(BaseModel):
    like_count: int
    is_liked: bool


class ReportCreateRequest(BaseModel):
    target_type: ReportTargetType
    target_id: int
    reason: ReportReason


# ── Quiz DTOs ──────────────────────────────────────────────────────────────────


class QuizResponse(BaseSerializerModel):
    id: int
    question: str
    option_a: str
    option_b: str
    option_c: str
    option_d: str
    category: QuizCategory
    quiz_date: date | None


class QuizAnswerRequest(BaseModel):
    selected_option: QuizOption


class QuizAnswerResponse(BaseSerializerModel):
    is_correct: bool
    correct_option: QuizOption
    explanation: str
    points_earned: int


class QuizAttemptHistoryItem(BaseSerializerModel):
    quiz_id: int
    quiz_date: date | None
    question: str
    category: QuizCategory
    selected_option: QuizOption
    is_correct: bool
    points_earned: int
    attempted_at: datetime

File: web/src/types/community.ts
export type PostCategory = "NOTICE" | "INFO" | "FREE";

export interface PostListItem {
  id: number;
  title: string;
  category: PostCategory;
  is_pinned: boolean;
  view_count: number;
  comment_count: number;
  like_count: number;
  author_id: string;
  author_nickname: string | null;
  created_at: string;
}

export interface PostDetail extends PostListItem {
  content: string;
  updated_at: string;
  is_liked: boolean;
}

export interface PostListResponse {
  items: PostListItem[];
  total: number;
  page: number;
  size: number;
}

export interface PostCreateRequest {
  title: string;
  content: string;
  category: PostCategory;
}

export interface PostUpdateRequest {
  title?: string;
  content?: string;
  category?: PostCategory;
}

export interface Comment {
  id: number;
  content: string;
  author_id: string;
  author_nickname: string | null;
  parent_id: number | null;
  created_at: string;
  updated_at: string;
  like_count: number;
  is_liked: boolean;
  replies: Comment[];
}

export interface LikeResponse {
  like_count: number;
  is_liked: boolean;
}

export interface CommentCreateRequest {
  content: string;
  parent_id?: number;
}

export type ReportTargetType = "POST" | "COMMENT" | "VERIFICATION";
export type ReportReason = "ABUSE" | "MISINFORMATION" | "PRIVACY" | "AD" | "FRAUD" | "ETC";

export interface ReportCreateRequest {
  target_type: ReportTargetType;
  target_id: number;
  reason: ReportReason;
}

// ── Quiz ──────────────────────────────────────────────────────────────────────
export type QuizCategory = "BLOOD_SUGAR" | "BLOOD_PRESSURE" | "DIET" | "EXERCISE" | "GENERAL";
export type QuizOption = "A" | "B" | "C" | "D";

export interface QuizResponse {
  id: number;
  question: string;
  option_a: string;
  option_b: string;
  option_c: string;
  option_d: string;
  category: QuizCategory;
  quiz_date: string;
}

export interface TodayQuizResponse {
  quiz: QuizResponse;
  already_answered: boolean;
}

export interface QuizAnswerRequest {
  selected_option: QuizOption;
}

export interface QuizAnswerResponse {
  is_correct: boolean;
  correct_option: QuizOption;
  explanation: string;
  points_earned: number;
}

export interface QuizAttemptHistoryItem {
  quiz_id: number;
  quiz_date: string;
  question: string;
  category: QuizCategory;
  selected_option: QuizOption;
  is_correct: boolean;
  points_earned: number;
  attempted_at: string;
}
